packages/aws-cdk-lib/aws-cloudfront-origins/lib/vpc-origin.ts

```diff
@@ -0,0 +1,130 @@
+   *
+   * Note that values over 60 seconds are possible only after a limit increase request for the origin response timeout quota
+   * has been approved in the target account; otherwise, values over 60 seconds will produce an error at deploy time.
+   *
+   * @default Duration.seconds(5)
+   */
+  readonly keepaliveTimeout?: cdk.Duration;
+}
+
+/**
+ * Properties to define a VPC origin with endpoint.
+ */
+export interface VpcOriginWithEndpointProps extends VpcOriginProps, cloudfront.VpcOriginOptions {
+}
+
+/**
+ * Represents a distribution's VPC origin.
+ */
+export abstract class VpcOrigin extends cloudfront.OriginBase {
+  /**
+   * Create a VPC origin with an existing VPC origin resource.
+   */
+  public static withVpcOrigin(origin: cloudfront.IVpcOrigin, props?: VpcOriginProps): VpcOrigin {
+    return new VpcOriginWithVpcOrigin(origin, props);
+  }
+
+  /**
+   * Create a VPC origin with an EC2 instance.
+   */
+  public static withEc2Instance(instance: IInstance, props?: VpcOriginWithEndpointProps): VpcOrigin {
+    return new VpcOriginWithEndpoint(cloudfront.VpcOriginEndpoint.ec2Instance(instance), props);
+  }
+
+  /**
+   * Create a VPC origin with an Application Load Balancer.
+   */
+  public static withApplicationLoadBalancer(alb: IApplicationLoadBalancer, props?: VpcOriginWithEndpointProps): VpcOrigin {
+    return new VpcOriginWithEndpoint(cloudfront.VpcOriginEndpoint.applicationLoadBalancer(alb), props);
+  }
+
+  /**
+   * Create a VPC origin with a Network Load Balancer.
+   */
+  public static withNetworkLoadBalancer(nlb: INetworkLoadBalancer, props?: VpcOriginWithEndpointProps): VpcOrigin {
+    return new VpcOriginWithEndpoint(cloudfront.VpcOriginEndpoint.networkLoadBalancer(nlb), props);
+  }
+
+  protected vpcOrigin?: cloudfront.IVpcOrigin;
+
+  protected constructor(domainName: string, protected readonly props: VpcOriginProps) {
+    super(domainName, props);
+
+    validateSecondsInRangeOrUndefined('readTimeout', 1, 180, props.readTimeout);
+    validateSecondsInRangeOrUndefined('keepaliveTimeout', 1, 180, props.keepaliveTimeout);
+  }
+
+  protected renderVpcOriginConfig(): cloudfront.CfnDistribution.VpcOriginConfigProperty | undefined {
+    if (!this.vpcOrigin) {
+      throw new cdk.UnscopedValidationError('VPC origin cannot be undefined.');
+    }
+    return {
+      vpcOriginId: this.vpcOrigin.vpcOriginId,
+      originReadTimeout: this.props.readTimeout?.toSeconds(),
+      originKeepaliveTimeout: this.props.keepaliveTimeout?.toSeconds(),
+    };
+  }
+}
+
+class VpcOriginWithVpcOrigin extends VpcOrigin {
+  constructor(protected vpcOrigin: cloudfront.IVpcOrigin, props: VpcOriginProps = {}) {
+    const domainName = props.domainName ?? vpcOrigin.domainName;
+    if (!domainName) {
+      throw new cdk.UnscopedValidationError("'domainName' must be specified when no default domain name is defined.");
+    }
+    super(domainName, props);
+  }
+}
+
+class VpcOriginWithEndpoint extends VpcOrigin {
```

Comment:
what's the use of this class if it's not exported?

Reply:
Similar to `S3BucketOriginWithOAC` and `S3BucketOriginWithOAI`.
`VpcOriginWith****` is used internally by `VpcOrigin.with***` static methods.

`S3BucketOrigin.withOriginAccessControl` static method creates `S3BucketOriginWithOAC`.
https://github.com/aws/aws-cdk/blob/4b0094b38f066cbf84425c70da4ed6f328e95d6c/packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts#L70-L72
`S3BucketOriginWithOAC` is not exported.
https://github.com/aws/aws-cdk/blob/4b0094b38f066cbf84425c70da4ed6f328e95d6c/packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts#L108